From: Tomas Henzl <thenzl@redhat.com>
Date: Tue, 4 Jan 2011 14:23:00 -0500
Subject: [scsi] megaraid: give FW more time to recover from reset
Message-id: <4D232D44.8060607@redhat.com>
Patchwork-id: 31027
O-Subject: [RHEL5.6 PATCH] megaraid: give more time to FW to come out from failed
	status
Bugzilla: 665427
RH-Acked-by: Jiri Pirko <jpirko@redhat.com>
RH-Acked-by: Jarod Wilson <jarod@redhat.com>

This is for bz#665427

If error occurs during I/O, the SCSI driver resets the controller to
restore it to normal state.
But on RHEL5, the waiting time until reset completion in megaraid_sas
driver is too short. The driver incorrectly recognizes the controller
as stalled, and the system stalls.

This is a part of an upstream patch -
http://marc.info/?l=linux-scsi&m=129018442416761&w=2

The patch is reported to cure the issue, I haven't tested it myself.
It impacts a state when the firmware already is in a fault state
so I expect it will not harm normal operation.

Tomas

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/scsi/megaraid/megaraid_sas.c b/drivers/scsi/megaraid/megaraid_sas.c
index bdf9150..5555aa0 100644
--- a/drivers/scsi/megaraid/megaraid_sas.c
+++ b/drivers/scsi/megaraid/megaraid_sas.c
@@ -2610,7 +2610,9 @@ megasas_transition_to_ready(struct megasas_instance* instance)
 		case MFI_STATE_FAULT:
 
 			printk(KERN_DEBUG "megasas: FW in FAULT state!!\n");
-			return -ENODEV;
+			max_wait = MEGASAS_RESET_WAIT_TIME;
+			cur_state = MFI_STATE_FAULT;
+			break;
 
 		case MFI_STATE_WAIT_HANDSHAKE:
 			/*
